test/e2e/storage/testsuites/volumes.go

```diff
@@ -89,101 +89,76 @@ func skipExecTest(driver TestDriver) {
 	}
 }
 
-func createVolumesTestInput(pattern testpatterns.TestPattern, resource genericVolumeTestResource) volumesTestInput {
-	var fsGroup *int64
-	driver := resource.driver
-	dInfo := driver.GetDriverInfo()
-	f := dInfo.Config.Framework
-	volSource := resource.volSource
+func (t *volumesTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	var (
+		dInfo       = driver.GetDriverInfo()
+		config      *PerTestConfig
+		testCleanup func()
+		resource    *genericVolumeTestResource
+	)
 
-	if volSource == nil {
-		framework.Skipf("Driver %q does not define volumeSource - skipping", dInfo.Name)
+	// No preconditions to test. Normally they would be in a BeforeEach here.
+
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("volumeio")
+
+	init := func() {
+		// Now do the more expensive test initialization.
+		config, testCleanup = driver.PrepareTest(f)
+		resource = createGenericVolumeTestResource(driver, config, pattern)
+		if resource.volSource == nil {
```

Comment:
same

Reply:
Same :man_shrugging:. I wanted to avoid changing test logic.